catboost/libs/metrics/metric.cpp

```diff
@@ -2086,6 +2086,122 @@ void TAverageGain::GetBestValue(EMetricBestValue* valueType, float*) const {
+
+    auto metric = new TNormalizedGINIMetric();
+    metric->PositiveClass = positiveClass;
+    metric->IsMultiClass = true;
+    return metric;
+}
+
+TMetricHolder TNormalizedGINIMetric::Eval(
+        const TVector<TVector<double>>& approx,
+        const TVector<float>& target,
+        const TVector<float>& weightIn,
+        const TVector<TQueryInfo>& /*queriesInfo*/,
+        int begin,
+        int end,
+        NPar::TLocalExecutor& /* executor */
+) const {
+    Y_ASSERT((approx.size() > 1) == IsMultiClass);
+    const auto& approxVec = approx.ysize() == 1 ? approx.front() : approx[PositiveClass];
+    Y_ASSERT(approxVec.size() == target.size());
+    const auto& weight = UseWeights ? weightIn : TVector<float>{};
+
+    TVector<double> approxCopy(approxVec.begin() + begin, approxVec.begin() + end);
+    TVector<double> targetCopy(target.begin() + begin, target.begin() + end);
+
+    if (!IsMultiClass) {
+        for (ui32 i = 0; i < targetCopy.size(); ++i) {
+            targetCopy[i] = targetCopy[i] > Border;
+        }
+    }
+
+    if (approx.ysize() > 1) {
+        int positiveClass = PositiveClass;
+        ForEach(targetCopy.begin(), targetCopy.end(), [positiveClass](double& x) {
+            x = (x == static_cast<double>(positiveClass));
+        });
+    }
+
+    TVector<NMetrics::TSample> samples;
+    if (weight.empty()) {
+        samples = NMetrics::TSample::FromVectors(targetCopy, approxCopy);
+    } else {
+        TVector<double> weightCopy(weight.begin() + begin, weight.begin() + end);
+        samples = NMetrics::TSample::FromVectors(targetCopy, approxCopy, weightCopy);
+    }
+
+    double normalizedGini = 2 * CalcAUC(&samples) - 1;
+
+    TMetricHolder error(2);
+    error.Stats[0] = normalizedGini;
+    error.Stats[1] = 1.0;
+    return error;
+}
+
+TString TNormalizedGINIMetric::GetDescription() const {
+    if (IsMultiClass) {
+        const TMetricParam<int> positiveClass("class", PositiveClass, /*userDefined*/true);
+        return BuildDescription(ELossFunction::NormalizedGINI, UseWeights, positiveClass);
+    } else {
+        return BuildDescription(ELossFunction::NormalizedGINI, UseWeights, MakeBorderParam(Border));
+    }
+}
+
+void TNormalizedGINIMetric::GetBestValue(EMetricBestValue* valueType, float*) const {
+    *valueType = EMetricBestValue::Max;
+}
+
+
+/* FairLoss */
+
+TFairLossMetric::TFairLossMetric(double c)
+        : c(c) {
+}
+
+TMetricHolder TFairLossMetric::EvalSingleThread(
+        const TVector<TVector<double>>& approx,
+        const TVector<float>& target,
+        const TVector<float>& weight,
+        const TVector<TQueryInfo>& /*queriesInfo*/,
+        int begin,
+        int end
+) const {
+    CB_ENSURE(approx.size() == 1, "Metric FairLoss supports only single-dimensional data");
+
+    const auto& approxVec = approx.front();
+    Y_ASSERT(approxVec.size() == target.size());
+
+    TMetricHolder error(2);
+    for (int k = begin; k < end; ++k) {
```

Comment:
Please use meaningful identifiers instead of k, w and x, e.g. idx, documentWeight, targetMismatch.